src/Chainweb/Pact/Backend/PactState.hs

```diff
@@ -0,0 +1,345 @@
+      deltaToObject = \case
+        Old x -> Aeson.object
+          [ "old" .= x
+          ]
+        New x -> Aeson.object
+          [ "new" .= x
+          ]
+        Delta x y -> Aeson.object
+          [ "old" .= x
+          , "new" .= y
+          ]
+        Same _ -> Aeson.Null
+
+data UserTable = UserTable
+  { tableName :: !Text
+  , rows :: [PactRow]
+  }
+  deriving stock (Eq, Ord, Show)
+
+data PactRow = PactRow
+  { rowKey :: !ByteString
+  , rowData :: !ByteString
+  , txId :: !Int64
+  }
+  deriving stock (Eq, Ord, Show)
+
+instance ToJSON PactRow where
+  toJSON pr = Aeson.object
+    [ "row_key" .= Text.decodeUtf8 pr.rowKey
+    , "row_data" .= Text.decodeUtf8 pr.rowData
+    , "tx_id" .= pr.txId
+    ]
+
+getActiveRows :: UserTable -> UserTable
+getActiveRows (UserTable name rows) = UserTable name
+  $ List.map takeHead
+  $ List.map (List.sortOn (Down . txId))
+  $ List.groupBy (\x y -> rowKey x == rowKey y)
+  $ List.sortOn rowKey rows
+  where
+    takeHead :: [a] -> a
+    takeHead = \case
+      [] -> error "getLatestPactState.getActiveRows.takeHead: impossible case"
+      (x : _) -> x
+
+utf8ToText :: Utf8 -> Text
+utf8ToText (Utf8 u) = Text.decodeUtf8 u
+
+data Config = Config
+  { pactDbDir :: FilePath
+  , compactDir :: FilePath
+  , chainwebVersion :: ChainwebVersion
+  , logDir :: FilePath
+  , numThreads :: Int
+  }
+
+main :: IO ()
+main = do
+  cfg <- execParser opts
+
+  when (cfg.pactDbDir == cfg.compactDir) $ do
+    Text.putStrLn "Pact database directory and compacted Pact database directory cannot be the same."
+    exitFailure
+
+  cids <- do
+    -- Get the latest block height on chain 0 for the purpose of calculating all
+    -- the chain ids at the current (version,height) pair
+    latestBlockHeight <- C.withDefaultLogger Error $ \logger -> do
+      let resetDb = False
+      withSqliteDb (unsafeChainId 0) logger cfg.pactDbDir resetDb $ \(SQLiteEnv db _) -> do
+        getLatestBlockHeight db
+    pure $ List.sort $ F.toList $ chainIdsAt cfg.chainwebVersion latestBlockHeight
+
+  flip (pooledMapConcurrentlyN_ cfg.numThreads) cids $ \cid -> do
+    C.withPerChainFileLogger cfg.logDir cid Debug $ \logger' -> do
+      let logger = over setLoggerScope (("chain-id", sshow cid) :) logger'
+      let resetDb = False
+      withSqliteDb cid logger cfg.compactDir resetDb $ \(SQLiteEnv cpDb _) -> do
+        latestBlockHeight <- getLatestBlockHeight cpDb
+        void $ C.compact latestBlockHeight logger cpDb []
+
+        withSqliteDb cid logger cfg.pactDbDir resetDb $ \(SQLiteEnv db _) -> do
+          diffEmptyRef <- newIORef True
+
+          let diff = diffLatestPactState (getLatestPactState db) (getLatestPactState cpDb)
+          flip S.mapM_ diff $ \utd -> do
+            writeIORef diffEmptyRef False
+            loggerFunIO logger Warn $ toLogMessage $
+              TextLog $ Text.decodeUtf8 $ BSL.toStrict $ Aeson.encode utd
+
+          diffEmpty <- readIORef diffEmptyRef
```

Comment:
This should be something that happens at the very end. We should have an `IORef (Map ChainId Bool)`.